### PR TITLE
Implementation for init=

### PR DESCRIPTION
**Pull request by benharsh**
This PR implements support for the new kind of copy initializer, named "init=". This copy initializer will follow similar rules as a regular initializer (e.g. initialize fields in order). The init= method will be invoked for compiler-inserted copies of records, and for initialization of variables from an expression. A simple case is initializing from the same type:

```chpl
var orig : MyRecord;
var copy : MyRecord = orig; // copy.init=(orig)
```

A more interesting case is when initializing from a different type:

```chpl
record R {
  var x : int;

  proc init=(other : int) {
    this.x = other;
  }
}

var r : R = 5; // r.init=(5);
```

Generic types require an additional type formal that represents the intended instantiation:

```chpl
record R {
  type T;
  var x : T;

  proc init=(type ThisType, other : ThisType.T) {
    this.T = ThisType.T;
    this.x = other;
  }
}

var x : R(int) = 5;
var y : R(string) = "hello";
```

As a result of this functionality, atomic variables can be initialized from values:
```chpl
var x : atomic int = 5;
```

This PR allows for the old-style copy initializers to still be invoked, unless the user has implemented their own init= methods.

Testing:
- [x] local
- [x] memleaks
- [x] valgrind
- [x] gasnet

### CODE REVIEW

**Comment by benharsh** [comment]
This PR updates the "records.chpl" primer to use init=, but would pass if we used the old-style copy initializer. Do we want to make this change where users would see it?

File: compiler/resolution/functionResolution.cpp [hunk 1]
```diff
@@ -4833,71 +4841,65 @@ static void resolveInitVar(CallExpr* call) {
...
     if (targetType->symbol->hasFlag(FLAG_GENERIC)) {
-      Type* useType = getInstantiationType(srcType, targetType);
+      Type* inst = getInstantiationType(srcType, targetType);
 
-      if (useType == NULL) {
-        USR_FATAL(call, "Could not coerce %s to %s in initialization",
-                        srcType->symbol->name,
-                        targetType->symbol->name);
+      // Does not allow initializations of the form:
+      //   var x : MyGenericType = <expr>;
```
Comment (mppf) [inline compiler/resolution/functionResolution.cpp]: But we do allow this, in cases where getInstantiationType knows what to do, right?

E.g.

```
record R { type t; }
var x : R = new R(int);
```

Doesn't this still work after this PR? (If not, we'll have a lot of problems with owned and shared...)

File: compiler/resolution/functionResolution.cpp [hunk 2]
```diff
@@ -4833,71 +4841,65 @@ static void resolveInitVar(CallExpr* call) {
...
 
-  // Next, the primitive is replaced with appropriate AST (depending on the
-  // relevant types and flags).
+      CallExpr* coerce = new CallExpr(PRIM_COERCE,
+                                      srcExpr->copy(),
+                                      targetTypeExpr);
+      CallExpr* move = new CallExpr(PRIM_MOVE, tmp, coerce);
 
-  // Target type will never be a reference
-  targetType = targetType->getValType();
```
Comment (mppf) [inline compiler/resolution/functionResolution.cpp]: I'm curious if you can explain why you removed this line?
It looks like some (many?) of the uses of targetType later are getting the value type.

File: modules/internal/Atomics.chpl [hunk 3]
```diff
@@ -187,6 +187,21 @@ module Atomics {
       atomic_init(_v, default);
     }
 
+    pragma "no doc"
+    proc init=(other : AtomicBool) {
+      this._v = other._v;
+    }
+
+    pragma "no doc"
+    proc init=(other : bool) {
```
Comment (mppf) [inline modules/internal/Atomics.chpl]: I didn't see much PR message but this new atomics support would be worth calling out in the PR message

File: modules/internal/ChapelSyncvar.chpl [hunk 4]
```diff
@@ -134,6 +134,17 @@ module ChapelSyncvar {
       this.isOwned = false;
     }
 
+    proc init=(type ThisType, const other : _syncvar) {
+      if other.type != ThisType then compilerError("init= error");
+      this.init(other);
+    }
```
Comment (mppf) [inline modules/internal/ChapelSyncvar.chpl]: Are these actually used? It looks like the implementation was still calling initCopy for sync vars. Is initCopy calling init= now?

File: modules/internal/OwnedObject.chpl [hunk 5]
```diff
@@ -240,6 +240,17 @@ module OwnedObject {
        that takes over ownership from `src`. `src` will
        refer to `nil` after this call.
      */
+    proc init=(type ThisType, pragma "leaves arg nil" pragma "nil from arg" ref src:_owned) {
+      // Use 'ThisType.chpl_t' in case RHS is a subtype
+      this.chpl_t = ThisType.chpl_t;
+      this.chpl_p = src.release();
+    }
+
+    proc init=(type ThisType, src : _nilType) {
```
Comment (mppf) [inline modules/internal/OwnedObject.chpl]: This needs docs or a no doc

File: modules/internal/OwnedObject.chpl [hunk 6]
```diff
@@ -240,6 +240,17 @@ module OwnedObject {
...
+      this.init(ThisType.chpl_t);
+    }
+
+    // Copy-init implementation to allow for 'new _owned(foo)' in module code
```
Comment (mppf) [inline modules/internal/OwnedObject.chpl]: same here

Comment (mppf) [reply modules/internal/OwnedObject.chpl]: I'm not sure we want to continue calling this one though.... do you know where it is called?

File: modules/standard/BigInteger.chpl [hunk 7]
```diff
@@ -216,6 +220,10 @@ module BigInteger {
       this.localeId = chpl_nodeID;
     }
 
+    proc init=(num : integral) {
+      this.init(num);
+    }
```
Comment (mppf) [inline modules/standard/BigInteger.chpl]: 👍

File: test/classes/initializers/initequals/basic.chpl [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  var A : R;
+  var B = new R(5);
+  writeln(A);
+  writeln(B);
+}
+writeln("\n\n");
+
+writeln("----- User-Defined Copy Initialization -----");
+
+// will be 'init=' eventually
```
Comment (mppf) [inline test/classes/initializers/initequals/basic.chpl]: Maybe it's time to remove this comment?

File: compiler/AST/AggregateType.cpp [hunk 9]
```diff
@@ -1665,43 +1677,50 @@ void AggregateType::buildCopyInitializer() {
...
 
-    FnSymbol*  fn    = new FnSymbol("init");
+    bool isGeneric = false;
+    // If this type is generic, then the 'other' formal needs to be generic as
+    // well
+    for_fields(fieldDefExpr, this) {
+      if (VarSymbol* field = toVarSymbol(fieldDefExpr)) {
+        if (field->hasFlag(FLAG_SUPER_CLASS) == false) {
+          if (field->hasFlag(FLAG_PARAM) || field->isType() ||
+              (field->defPoint->init == NULL && field->defPoint->exprType == NULL)) {
```
[the remark this answers is not in the document]
Comment (benharsh) [reply compiler/AST/AggregateType.cpp]: I'm sure I tried using that at first, but it didn't work. I'll add a TODO.

File: compiler/include/flags_list.h [hunk 10]
```diff
@@ -110,7 +110,7 @@ symbolFlag( FLAG_DEFAULT_INTENT_IS_REF, ypr, "default intent is ref", "The defau
 // (ie. it is a ref if it is modified in the function body)
 symbolFlag( FLAG_DEFAULT_INTENT_IS_REF_MAYBE_CONST, ypr, "default intent is ref if modified", "The default intent for this type is ref if modified const ref otherwise")
 
-symbolFlag( FLAG_DEFAULT_COPY_INIT, npr, "default copy initializer", ncm )
+symbolFlag( FLAG_COPY_INIT, npr, "default copy initializer", ncm )
```
[the remark this answers is not in the document]
Comment (benharsh) [reply compiler/include/flags_list.h]: Just an oversight, I'll fix.

File: compiler/resolution/functionResolution.cpp [hunk 11]
```diff
@@ -3239,6 +3239,14 @@ static FnSymbol* resolveForwardedCall(CallInfo& info, bool checkOnly) {
     return NULL;
   }
 
+  if (call->isNamedAstr(astrInit) && call->numActuals() >= 1) {
+    if (SymExpr* se = toSymExpr(call->get(1))) {
+      if (se->symbol() == gMethodToken) {
+        return NULL;
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (benharsh) [reply compiler/resolution/functionResolution.cpp]: There was an issue at one point during development where an initializer was being forwarded. I'll add a comment.

File: compiler/resolution/functionResolution.cpp (hunk 1 above)
Comment (benharsh) [reply compiler/resolution/functionResolution.cpp]: We do allow it when the RHS is an instantiation of the LHS, I'll update the comment.

File: compiler/resolution/functionResolution.cpp (hunk 2 above)
Comment (benharsh) [reply compiler/resolution/functionResolution.cpp]: Some of those branches have different results if 'targetType' is a reference, so I couldn't throw that information away. I could create a simple 'targetTypeVal' variable if you want to minimize the number of "->getValType()" calls.

File: modules/internal/ChapelSyncvar.chpl (hunk 4 above)
Comment (benharsh) [reply modules/internal/ChapelSyncvar.chpl]: good catch, these are left over from an earlier effort to use init=, but I backed away from it when it sort of blew up. I'll remove these and the _singlevar one.

File: modules/internal/OwnedObject.chpl (hunk 6 above)
Comment (benharsh) [reply modules/internal/OwnedObject.chpl]: I think there are some 'new _owned' in this module, not sure about other modules though.

File: compiler/passes/normalize.cpp [hunk 12]
```diff
@@ -155,7 +155,7 @@ void normalize() {
         }
         USR_FATAL_CONT(fn, "Type '%s' defines a constructor here", ct->symbol->name);
 
-      } else if (fn->isInitializer() == true) {
+      } else if (fn->isInitializer() || fn->isCopyInit()) {
```
[the remark this answers is not in the document]
Comment (benharsh) [reply compiler/passes/normalize.cpp]: I would like something like this too, and something that would handle the strcmp cases too. Do you want that to be done for this PR?

[the remark this answers is not in the document]
Comment (mppf) [reply compiler/passes/normalize.cpp]: Only if you think it'll help you get the code correct. If you choose to put it off, let's make an issue and put it to get worked on soon.

**Pull request merged**